src/backend/distributed/commands/multi_copy.c

```diff
@@ -1853,9 +2116,19 @@ CitusCopyDestReceiverStartup(DestReceiver *dest, int operation,
 	copyOutState->rowcontext = GetPerTupleMemoryContext(copyDest->executorState);
 	copyDest->copyOutState = copyOutState;
 
-	/* prepare output functions */
-	copyDest->columnOutputFunctions =
-		ColumnOutputFunctions(inputTupleDescriptor, copyOutState->binary);
+	/* prepare functions to call on received tuples */
+	{
+		TupleDesc destTupleDescriptor = distributedRelation->rd_att;
+		int columnCount = inputTupleDescriptor->natts;
+		Oid *finalTypeArray = palloc0(columnCount * sizeof(Oid));
+
+		copyDest->columnCoercionPaths =
+			ColumnCoercionPaths(destTupleDescriptor, inputTupleDescriptor,
+								tableId, columnNameList, finalTypeArray);
```

Comment:
Is it worth skipping this for regular COPY commands?

Reply:
That would make sense, how would you distinguish between the two cases? I don't think it makes a big difference, presumably the types will always match, we'll get a lot of `RELABEL_TYPE`, and receiving isn't slowed down at all, despite startup doing a little more work.